designs/2020-rule-tester-debugger/README.md

```diff
@@ -0,0 +1,78 @@
+- Repo: eslint/eslint
+- Start Date: 2020-09-21
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: [CoryDanielson](https://github.com/corydanielson)
+
+# `RuleTester` debugger improvements
+
+## Summary
+
+Add a `before` hook to the valid/invalid code configuration, so that developers can inject a debugger into RuleTester before code is passed into the rule being tested.
+
+## Motivation
+
+Giving developers a mechanism to add a debugger before specific code is passed into their rule, would streamline and improve the developer experience when writing rules. Currently, when a developer wants to debug their rule against specific test code they will most likely do one of the following:
+    1. Comment out the other test code blocks - the developer will be unable to see if they have broken an existing test while changing code.
+    2. Rearrange the test code blocks so that the one they wish to debug comes first - this discourages or breaks organization
+    3. Press the "Play/Resume" button while debugging X times until the debugger is paused at the code block they want to debug - tedious, especially when there are lots of tests or add new ones (X may change)
+    4. Some clever conditional debugging - the developer is spending time writing conditional debugger logic instead of the eslint rule.
+
+## Detailed Design
+
+1. Add an optional `before` function property to the config for valid/invalid code.
+2. Update RuleTester to call this function (if it exists) just prior to the valid/invalid code being passed into the rule.
```

Comment:
`RuleTester` doesn't call rules directly. It calls [`linter.verify(...)`](https://github.com/eslint/eslint/blob/b5e011c865e95d700d29cb9a4ba71c671d99e423/lib/rule-tester/rule-tester.js#L556). If `RuleTester` were to call `before` immediately before the `verify()` call, it would still take developers on a tour of ESLint internals before they get to their rule code.

Reply:
That might make this a non-starter. A way to avoid stepping through the internals would be if the rule was conditionally edited and a debugger was inserted into it by RuleTester